src/librbd/librbd.cc

```diff
@@ -61,6 +62,16 @@ namespace {
 
 TracepointProvider::Traits tracepoint_traits("librbd_tp.so", "rbd_tracing");
 
+buffer::raw* create_write_raw(librbd::ImageCtx *ictx, const char *buf,
+                              size_t len) {
+  // cannot allow the cache (if enabled) to retain reference to user memory
+  if (ictx->cache) {
+    return buffer::copy(buf, len);
+  } else {
+    return buffer::create_static(len, const_cast<char*>(buf));
```

Comment:
indeed -- it was Sage's example of a PG mapping update and the messenger threads racing between old and new PGs. In a perfect world, I would expect that once librados ACKs, the memory wouldn't be used again. Baring that, the C API just a needs to always copy at least once.